app/src/main/java/com/example/travelapp/BinaryMinHeap.java

```diff
@@ -0,0 +1,166 @@
+package com.example.travelapp;
+
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * BinaryMinHeap class. Implemented with an ArrayList that can be
+ * viewed as a nearly complete binary tree, where every node in the
+ * tree corresponds to an element in the array. Taken and adapted
+ * from a Penn homework assignment.
+ * @param <Key> - The key used to determine priority in the heap.
+ *             Does not have to be distinct.
+ * @param <V> - The value of what we are adding to the heap. In
+ *           Dijkstras, this would correspond to the node number.
+ */
+
+public class BinaryMinHeap<Key extends Comparable<Key>, V> {
+
+    // arrayList can be viewed as a nearly complete binary tree
+    private ArrayList<Entry<Key, V>> arrayList;
+    // indices is a HashMap used for easier lookup of where values
+    // are in the array
+    private HashMap<V, Integer> indices;
+
+    public BinaryMinHeap() {
+        arrayList = new ArrayList<Entry<Key, V>>();
+        indices = new HashMap<V, Integer>();
+    }
+
+    public int size() {
+        return arrayList.size();
+    }
+
+    public boolean containsValue(V value) {
+        return indices.containsKey(value);
+    }
+
+    public void add(Key key, V value) {
+        if (key == null) {
+            throw new IllegalArgumentException("key is null!");
+        }
+        if (indices.containsKey(value)) {
+            throw new IllegalArgumentException("value is already in the min-heap!");
+        }
+
+        arrayList.add(new Entry<Key, V>(key, value));
+        // set index to where we just added the Entry (at the end = index size - 1 of the arrayList)
+        int index = arrayList.size() - 1;
+        // calculates the index of the parent of the current node, given the current node's index
+        Function<Integer, Integer> calculateParentIndex = (n) -> { return (n + 1) / 2 - 1; };
```

Comment:
I like that this is it's own thing, but can this be moved outside of this function?